M58 release notes (#385)
pulumi/pulumi-hugo: themes/default/content/blog/pulumi-release-notes-m58/index.md

      - run: go mod download
      - uses: pulumi/actions@v3
        with:
          command: preview
          stack-name: dev
        env:
          PULUMI_ACCESS_TOKEN: ${{ secrets.PULUMI_ACCESS_TOKEN }}
```

{{% /choosable %}}
{{% choosable language csharp %}}

```yaml
name: Pulumi
on:
  - pull_request
jobs:
  preview:
    name: Preview
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-dotnet@v1
        with:
          dotnet-version: 3.1.x
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-region: ${{ secrets.AWS_REGION }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
      - uses: pulumi/actions@v3
        with:
          command: preview
          stack-name: dev
        env:
          PULUMI_ACCESS_TOKEN: ${{ secrets.PULUMI_ACCESS_TOKEN }}
```

{{% /choosable %}}

{{< /chooser >}}

Learn more in the [GitHub Actions User Guide]({{< relref "/docs/guides/continuous-delivery/github-actions" >}})

### Kubernetes: new `skipAwait` option

> For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API or `kubectl` experience. Previously, you had to do this by marking individual entities with a custom annotation to skip the await logic. Now, you can use the new `skipAwait` option on Helm charts and YAML components to turn off the await logic for those charts/components and all of their child resources. Please note: `skipAwait` has some downsides: outputs are not populated and incorrect deployments aren't identified. However, in cases where you don't need the result of the chart/component, you can speed up deployments using `skipAwait`.

Reviewer: ```suggestion
For some Kubernetes use cases, including Helm charts, you may want to turn off all await logic, so that using Kubernetes via Pulumi is similar to the default Kubernetes API or `kubectl` experience. Previously, this was accomplished by marking individual entities with a custom annotation to skip the await logic. The new `skipAwait` option on Helm charts and YAML components turnd off the await logic for those charts/components and all of their child resources. Please note: `skipAwait` changes the default behavior: outputs are not populated and incorrect deployments aren't identified. However, in cases where you don't need the result of the chart/component, you can accerlerate deployments using `skipAwait`.
```

Author: ```suggestion
For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that you can use Kubernetes via Pulumi in a way that feels more similar to the default Kubernetes API and `kubectl` experience. Previously, you could achieve this by marking individual entities with a custom annotation to skip the await logic. Now, you can use the new `skipAwait` option on Helm charts and YAML components to turn off the await logic for those charts/components and all of their child resources. Please note that `skipAwait` changes some default Pulumi behaviors: outputs are not populated and incorrect deployments aren't identified. However, in cases where you don't need the result of the chart/component, you can speed up deployments using `skipAwait`.
```

[See the full list](https://github.com/pulumi/pulumi-azure-native/blob/master/CHANGELOG.md#1150-2020-07-01)

### Google Native provider adds support for enums and more

The preview of the Google Native provider continues to evolve. We shipped 2 new versions of the Google Native provider (0.3.0 and 0.4.0) that include:

- Refined, shorter resource names that are easier to type and easier to find
- Enum support for all properties exposed as enums in the underlying Google Discovery Documents

[See the full list of updates](https://github.com/pulumi/pulumi-google-native/blob/master/CHANGELOG.md#040-2021-06-16)

## Pulumi CLI and core technologies

In this milestone, we shipped Pulumi versions [3.5.1](https://github.com/pulumi/pulumi/releases/tag/v3.5.1) and [3.6.0](https://github.com/pulumi/pulumi/releases/tag/v3.6.0). The full list of changes in each version is available in the linked changelog; read on to learn about some of the biggest changes.

### New `replaceOnChanges` resource option

> Sometimes, you may want to force the replacement of a resource even if the changes you've made to that resource don't indicate that they require a replacement. One example is Kubernetes `CustomResourceDefinition`s: many CRDs can't be updated in place due to forward-incompatibilities, so you may need to force a replacement. Previously, you could do this by changing the physical name of the resource, but there are cases where it is important to keep the physical name the same.

Reviewer: ```suggestion
Sometimes, you may want to  replace a resource even if the changes to the resource don't indicate that they require a replacement. One example is Kubernetes `CustomResourceDefinition`s: many CRDs can't be updated in place due to forward-incompatibilities, so you may need to force a replacement. Previously, this was accomplished by changing the physical name of the resource, but there are cases where it is important to keep the physical name the same.
```

> There are times when it is important to force a replacement of a resource even if changes to the resource do not indicate they require replacement. One example is a Kubernetes `CustomResourceDefinition`, none of whose properties force replacement, even though for many CRDs, they can't actually be updated in place (because of forward-incompatibilites). While it was previously possible to work around the lack of an option to force a replace by changing the physical name of the resource, there are cases where it is important to keep the physical name the same.

Reviewer 2: the 2nd and 3rd sentences are a little hard to follow (particularly the 2nd sentence).

> There are times when it is important to force a replacement of a resource even if changes to the resource do not indicate they require replacement. One example is a Kubernetes `CustomResourceDefinition`, none of whose properties force replacement, even though for many CRDs, they can't actually be updated in place (because of forward-incompatibilites). While it was previously possible to work around the lack of an option to force a replace by changing the physical name of the resource, there are cases where it is important to keep the physical name the same.

Author: ```suggestion
Sometimes, you may want to force the replacement of a resource even if the changes you've made to that resource don't indicate that they require a replacement. One example is Kubernetes `CustomResourceDefinition`s: many CRDs can't be updated in place due to forward-incompatibilities, so you may need to force a replacement. Previously, you could do this by changing the physical name of the resource, but there are cases where it is important to keep the physical name the same.
```

> Now, you can use the new `replaceOnChange` resource option to force a resource to be replaced when the specified properties are changed. This can be combined with the existing `deleteBeforeReplace` resource option to force a resource to be deleted before it's replaced. For example, you can specify that an AWS S3 Bucket resource be replaced every time its index document changes:

Reviewer: ```suggestion
You can use the new `replaceOnChange` resource option to force resource replacement when the specified properties are changed. This can be combined with the existing `deleteBeforeReplace` resource option to force a resource to be deleted before replacement. For example, you can specify that an AWS S3 Bucket resource be replaced every time its index document changes:
```

```typescript
const bucket = new aws.s3.Bucket("my-bucket", {
    website: {
        indexDocument: "index2.html",
    },
}, { replaceOnChanges: ["website.indexDocument"] });
```

Learn more at [pulumi/pulumi#7226](https://github.com/pulumi/pulumi/pull/7226)

### `pulumi watch` now supports a `--path` option

> Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This could cause duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths that need to be watched.

Reviewer: ```suggestion
Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This caused duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths to watch.
```

```shell
pulumi watch --path path/to/pulumi/code/
```

Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)

### Improved Python performance

> We've improved the performance of common Pulumi commands like `up` and `preview` for Python programs by removing some unneeded logging and version checks.

Reviewer: ```suggestion
We've improved the performance of common Pulumi commands like `up` and `preview` for Python programs by removing unnecessary logging and version checks.
```

```typescript
const bucket = new aws.s3.Bucket("my-bucket", {
    website: {
        indexDocument: "index2.html",
    },
}, { replaceOnChanges: ["website.indexDocument"] });
```

Learn more at [pulumi/pulumi#7226](https://github.com/pulumi/pulumi/pull/7226)

### `pulumi watch` now supports a `--path` option

Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This could cause duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths that need to be watched.

```shell
pulumi watch --path path/to/pulumi/code/
```

Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)

> ### Automation API: support for custom loggers in .NET
> 
> Many mature projects already have a preferred tool for logging events and exceptions. Previously, Automation API only supported a built-in logging tool, and it wasn't possible to use a project's preferred tool. Now, if you're using a .NET language like C#, you can supply your own logging tool, like Serilog.
> 
> Learn more at [pulumi/pulumi#7116](https://github.com/pulumi/pulumi/issues/7116)

Author: ```suggestion
```

---

> title: "July 7 release notes: updated pricing, watch paths, and more"

Reviewer 2: new pricing

Author: ```suggestion
title: "July 7 release notes: new pricing, watch paths, and more"
```

      - run: go mod download
      - uses: pulumi/actions@v3
        with:
          command: preview
          stack-name: dev
        env:
          PULUMI_ACCESS_TOKEN: ${{ secrets.PULUMI_ACCESS_TOKEN }}
```

{{% /choosable %}}
{{% choosable language csharp %}}

```yaml
name: Pulumi
on:
  - pull_request
jobs:
  preview:
    name: Preview
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-dotnet@v1
        with:
          dotnet-version: 3.1.x
      - name: Configure AWS Credentials
        uses: aws-actions/configure-aws-credentials@v1
        with:
          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
          aws-region: ${{ secrets.AWS_REGION }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
      - uses: pulumi/actions@v3
        with:
          command: preview
          stack-name: dev
        env:
          PULUMI_ACCESS_TOKEN: ${{ secrets.PULUMI_ACCESS_TOKEN }}
```

{{% /choosable %}}

{{< /chooser >}}

Learn more in the [GitHub Actions User Guide]({{< relref "/docs/guides/continuous-delivery/github-actions" >}})

### Kubernetes: new `skipAwait` option

> For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but for many use cases is the initially "expected" behavior. Now, with the `skipAwait` setting, you can opt out of await logic for an entire Pulumi Kubernetes program.

Reviewer 2: is it too opinionated to say "this isn't right in all cases"?

Author: Yeah, I wouldn't mind a phrasing check/assist from @lukehoban here.

> For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but for many use cases is the initially "expected" behavior. Now, with the `skipAwait` setting, you can opt out of await logic for an entire Pulumi Kubernetes program.

Reviewer 3: ```suggestion
For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API or `kubectl` experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but in some cases where the result of an operation is not material to the Pulumi program, this can speed up deployments. While we already supported the ability to mark individual entities with a custom annotation to skip await logic, we now support a new `skipAwait` option on Helm charts and yaml components that opts-out all resources in the component from being awaited on.
```

- Enum support for all properties exposed as enums in the underlying Google Discovery Documents

[See the full list of updates](https://github.com/pulumi/pulumi-google-native/blob/master/CHANGELOG.md#040-2021-06-16)

## Pulumi CLI and core technologies

In this milestone, we shipped Pulumi versions [3.5.1](https://github.com/pulumi/pulumi/releases/tag/v3.5.1) and [3.6.0](https://github.com/pulumi/pulumi/releases/tag/v3.6.0). The full list of changes in each version is available in the linked changelog; read on to learn about some of the biggest changes.

### `pulumi watch` now supports a `--path` option

Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This could cause duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths that need to be watched.

```shell
pulumi watch --path path/to/pulumi/code/
```

Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)

### Automation API: support for custom loggers in .NET

> Many mature projects already have a preferred tool for logging events and exceptions, but previously, Automation API only supported a built-in logger. Now, if you're using a .NET language like C#, you can supply your own logger, like Serilog.

Reviewer 2: the first sentence reads a little strangely, though I get what it's intending to say

Author: ```suggestion
Many mature projects already have a preferred tool for logging events and exceptions. Previously, Automation API only supported a built-in logging tool, and it wasn't possible to use a project's preferred tool. Now, if you're using a .NET language like C#, you can supply your own logging tool, like Serilog.
```

---
title: "July 7 releases: new pricing, replaceOnChanges, and more"
date: 2021-07-07T08:00:00-07:00
draft: false

> meta_desc: "Confluent Cloud, new `skipAwait` in K8s, new `replaceOnChanges` and `deleteBeforeReplace` options, restrict `pulumi watch` to paths, and usage-based pricing"

Author: ```suggestion
meta_desc: "Confluent Cloud, new `skipAwait` in K8s, new `replaceOnChanges` option, restrict `pulumi watch` to paths, and usage-based pricing"
```

meta_image: meta.png
authors:
    - alex-mullans
tags:
    - features
---

In this update:

- New and updated cloud providers
  - [Pulumi supports Confluent Cloud]({{< relref "/blog/pulumi-release-notes-m58#pulumi-supports-confluent-cloud" >}})
  - [GitHub Actions guide in all languages]({{< relref "/blog/pulumi-release-notes-m58#github-actions-guide-in-all-languages" >}})
  - [Kubernetes: new `skipAwait` option]({{< relref "/blog/pulumi-release-notes-m58#kubernetes-new-skipawait-option" >}})
  - [10 new resources in the Azure Native provider]({{< relref "/blog/pulumi-release-notes-m58#10-new-resources-in-the-azure-native-provider" >}})
  - [Google Native provider adds support for enums and more]({{< relref "/blog/pulumi-release-notes-m58#google-native-provider-adds-support-for-enums-and-more" >}})
- Pulumi CLI and core technologies

>   - [New `replaceOnChanges` and `deleteBeforeReplace` resource options]({{< relref "/blog/pulumi-release-notes-m58#new-replaceonchanges-and-deletebeforereplace-resource-options" >}})

Author: ```suggestion
  - [New `replaceOnChanges` resource option]({{< relref "/blog/pulumi-release-notes-m58#new-replaceonchanges-resource-option" >}})
```

          aws-region: ${{ secrets.AWS_REGION }}
          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
      - uses: pulumi/actions@v3
        with:
          command: preview
          stack-name: dev
        env:
          PULUMI_ACCESS_TOKEN: ${{ secrets.PULUMI_ACCESS_TOKEN }}
```

{{% /choosable %}}

{{< /chooser >}}

Learn more in the [GitHub Actions User Guide]({{< relref "/docs/guides/continuous-delivery/github-actions" >}})

### Kubernetes: new `skipAwait` option

For some Kubernetes use cases, including Helm charts, you may want to be able to turn off all await logic, so that using Kubernetes via Pulumi feels closer to the default Kubernetes API experience. This isn't right in all cases and has some downsides (outputs are not populated and incorrect deployments aren't identified), but for many use cases is the initially "expected" behavior. Now, with the `skipAwait` setting, you can opt out of await logic for an entire Pulumi Kubernetes program.

Learn more at [pulumi/pulumi-kubernetes#1421](https://github.com/pulumi/pulumi-kubernetes/issues/1421)

### 10 new resources in the Azure Native provider

> We shipped 5 new versions of the Azure Native provider (1.11.0 through 1.15.0) that collectively added 10 new resources that you can manage with the Azure Native provider.

Reviewer 4: Might be nice to add "including A, B, C" if that is possible?

[See the full list](https://github.com/pulumi/pulumi-azure-native/blob/master/CHANGELOG.md#1150-2020-07-01)

### Google Native provider adds support for enums and more

The preview of the Google Native provider continues to evolve. We shipped 2 new versions of the Google Native provider (0.3.0 and 0.4.0) that include:

- Refined, shorter resource names that are easier to type and easier to find
- Enum support for all properties exposed as enums in the underlying Google Discovery Documents

[See the full list of updates](https://github.com/pulumi/pulumi-google-native/blob/master/CHANGELOG.md#040-2021-06-16)

## Pulumi CLI and core technologies

In this milestone, we shipped Pulumi versions [3.5.1](https://github.com/pulumi/pulumi/releases/tag/v3.5.1) and [3.6.0](https://github.com/pulumi/pulumi/releases/tag/v3.6.0). The full list of changes in each version is available in the linked changelog; read on to learn about some of the biggest changes.

> ### New `replaceOnChanges` and `deleteBeforeReplace` resource options

Author: ```suggestion
### New `replaceOnChanges` resource option
```

```typescript
const bucket = new aws.s3.Bucket("my-bucket", {
    website: {
        indexDocument: "index2.html",
    },
}, { replaceOnChanges: ["website.indexDocument"], deleteBeforeReplace: true });
```

Learn more at [pulumi/pulumi#7226](https://github.com/pulumi/pulumi/pull/7226)

### `pulumi watch` now supports a `--path` option

Previously, the `pulumi watch` command always watched the entire Pulumi project directory (recursively). This could cause duplicate updates for developers working in monorepos or developers with application code and infrastructure code in the same directory. Now, you can add a `--path` option to the `pulumi watch` command and specify the exact paths that need to be watched.

```shell
pulumi watch --path path/to/pulumi/code/
```

Learn more at [pulumi/pulumi#6620](https://github.com/pulumi/pulumi/issues/6620)

### Automation API: support for custom loggers in .NET

> Many mature projects already have a preferred tool for logging events and exceptions. Previously, Automation API only supported a built-in logging tool, and it wasn't possible to use a project's preferred tool. Now, if you're using a .NET language like C#, you can supply your own logging tool, like Serilog.

Reviewer 4: I'm not sure this is an accurate description, but I'm also not entirely sure I follow what the improvement was here.  cc @t0yv0 for input to clarify what this should say (or whether this is small enough we should leave it out).

Reviewer 5: It's a community PR permitting using and customizing Serilog which enables a nice IDE integration for tests; I think it's small enough to leave out of notes here. If we want to include I can dig up the relevant details tomorrow.

Author: @t0yv0 thanks, I'll remove this item

Learn more at [pulumi/pulumi#7116](https://github.com/pulumi/pulumi/issues/7116)

> ### Automation API: more helpful error messages
> 
> Currently, it's not possible to use Automation API in a project that uses multiple versions of the `@pulumi/pulumi` package. We plan to fix this issue (tracked in [pulumi/pulumi#5449](https://github.com/pulumi/pulumi/issues/5449)) later this year. For now, to make it easier to troubleshoot and work around the issue, we've updated Automation API to fail fast when multiple versions of `@pulumi/pulumi` are detected and provide a helpful error message:
> 
> ```shell
> pulumi:pulumi:Stack (inlineNode-dev):
>     error: Unhandled exception: Error: Detected multiple versions of '@pulumi/pulumi' in use in an inline automation api program
>     Use the yarn 'resolutions' field to resolve this: https://github.com/pulumi/pulumi/issues/5449.
> ```
> 
> Learn more at [pulumi/pulumi#7349](https://github.com/pulumi/pulumi/issues/7333)
> 

Author: ```suggestion
```

### Improved Python performance

We've improved the performance of common Pulumi commands like `up` and `preview` for Python programs by removing some unneeded logging and version checks.

Learn more at [pulumi/pulumi#7291](https://github.com/pulumi/pulumi/issues/7291) and [pulumi/pulumi#7292](https://github.com/pulumi/pulumi/issues/7292)

### Pulumi Packages: TODO TYPES

> TODO

Reviewer 4: I think there was nothing new in this area in M58?

## Pulumi Service and Pulumi.com

### New usage-based pricing for your whole team

We launched Pulumi Team Edition, a new way for teams to adopt and use Pulumi and the Pulumi Service to collaborate on building, managing, and deploying cloud infrastructure as code. Pulumi Team Edition is priced based on the number of cloud resources under management, with a generous free tier to ensure that teams can get up and running with Pulumi Team Edition at no cost.

![Screenshot of new pricing tiers from https://pulumi.com/pricing](pricing-page.png)

> Learn more in the [announcement blog post]({{< relref "/blog/announcing-new-usage-based-pricing-for-your-whole-team" >}})

Reviewer 6: ```suggestion
Learn more in the [pricing announcement blog post]({{< relref "/blog/announcing-new-usage-based-pricing-for-your-whole-team" >}})
```
